ion/decimal.go

```diff
@@ -256,6 +264,19 @@ func (d *Decimal) Trunc() (int64, error) {
 	return strconv.ParseInt(str[:want], 10, 64)
 }
 
+// Round attempts to truncate this decimal to an int64, rounding any fractional bits.
+func (d *Decimal) Round() (int64, error) {
```

Comment:
Based on the name `Round`, I would expect this method to return the nearest integer value as a`Decimal`. (This is analogous to the behavior of the `Truncate` method below, which returns a `Decimal`.) Truncating the value to fit in the `int64` range would be pretty surprising, as that could lead to substantial data loss. What's the rationale behind changing it to an `int64`?

Reply:
This function is analogous to `Trunc` function which returns `int64`. `int64` will be sufficient because `readNsecs` function will shift decimal to 9 decimal places to the left before calling this function.

`func (b *bitstream) readNsecs(len uint64) (int, bool, error) {
	d, err := b.readDecimal(len)
	if err != nil {
		return 0, false, err
	}

	nsec, err := d.ShiftL(9).Trunc()
	if err != nil || nsec < 0 || nsec > 999999999 {
		msg := fmt.Sprintf("invalid timestamp fraction: %v", d)
		return 0, false, &SyntaxError{msg, b.pos}
	}

	nsec, err = d.ShiftL(9).Round()
	if err != nil {
		msg := fmt.Sprintf("invalid timestamp fraction: %v", d)
		return 0, false, &SyntaxError{msg, b.pos}
	}

	// Overflow to second.
	if nsec == 1000000000 {
		return 0, true, nil
	}

	return int(nsec), false, nil
}`

Reply:
> This function is analogous to Trunc function which returns int64

I hadn't noticed that; I would also change `Trunc` 😛 .

The problem is that while `int64` is suitable for *this* use case, we're making these functions part of `Decimal`'s public interface. Users working with a `Decimal` representation of their data will have varying requirements.

I would suggest either:
* making `Trunc` and `Round` private functions and change their names to reflect the data type conversion.
* making `Trunc` and `Round` return a `Decimal` (like `Truncate` does) and offering a separate method to convert to int64. This method would check for overflow and make sure the `Decimal` was an integer.

The latter is obviously more work; I'm fine with taking them private.

Reply:
Updated to private functions